src/vmm/src/devices/bus.rs

```diff
@@ -102,97 +106,69 @@ impl ConstantDevice {
 }
 
 impl BusDevice {
-    pub fn i8042_device_ref(&self) -> Option<&I8042Device> {
-        match self {
-            Self::I8042Device(x) => Some(x),
-            _ => None,
-        }
-    }
-    #[cfg(target_arch = "aarch64")]
-    pub fn rtc_device_ref(&self) -> Option<&RTCDevice> {
-        match self {
-            Self::RTCDevice(x) => Some(x),
-            _ => None,
-        }
-    }
-    pub fn boot_timer_ref(&self) -> Option<&BootTimer> {
-        match self {
-            Self::BootTimer(x) => Some(x),
-            _ => None,
-        }
-    }
-    pub fn mmio_transport_ref(&self) -> Option<&MmioTransport> {
-        match self {
-            Self::MmioTransport(x) => Some(x),
-            _ => None,
-        }
-    }
-    pub fn serial_ref(&self) -> Option<&SerialDevice<std::io::Stdin>> {
-        match self {
-            Self::Serial(x) => Some(x),
-            _ => None,
-        }
-    }
-
-    pub fn i8042_device_mut(&mut self) -> Option<&mut I8042Device> {
+    pub fn i8042_device_ref(&self) -> Option<MutexGuard<I8042Device>> {
         match self {
-            Self::I8042Device(x) => Some(x),
+            Self::I8042Device(x) => Some(x.lock().expect("Poisoned lock")),
             _ => None,
         }
     }
     #[cfg(target_arch = "aarch64")]
-    pub fn rtc_device_mut(&mut self) -> Option<&mut RTCDevice> {
+    pub fn rtc_device_ref(&self) -> Option<MutexGuard<RTCDevice>> {
         match self {
-            Self::RTCDevice(x) => Some(x),
+            Self::RTCDevice(x) => Some(x.lock().expect("Poisoned lock")),
             _ => None,
         }
     }
-    pub fn boot_timer_mut(&mut self) -> Option<&mut BootTimer> {
+    pub fn boot_timer_ref(&self) -> Option<MutexGuard<BootTimer>> {
         match self {
-            Self::BootTimer(x) => Some(x),
+            Self::BootTimer(x) => Some(x.lock().expect("Poisoned lock")),
             _ => None,
         }
     }
-    pub fn mmio_transport_mut(&mut self) -> Option<&mut MmioTransport> {
+    pub fn mmio_transport_ref(&self) -> Option<MutexGuard<MmioTransport>> {
         match self {
-            Self::MmioTransport(x) => Some(x),
+            Self::MmioTransport(x) => Some(x.lock().expect("Poisoned lock")),
             _ => None,
         }
     }
-    pub fn serial_mut(&mut self) -> Option<&mut SerialDevice<std::io::Stdin>> {
+    pub fn serial_ref(&self) -> Option<MutexGuard<SerialDevice<std::io::Stdin>>> {
         match self {
-            Self::Serial(x) => Some(x),
+            Self::Serial(x) => Some(x.lock().expect("Poisoned lock")),
             _ => None,
         }
     }
```

Comment:
Hmm. It looks like `serial_ref` is still in use for some aarch64 code, see `lib.rs:488`. The legacy device manager is x86 specific no? So although currently unused, should we not keep these functions available so that ARM machines can still access these devices?

Reply:
> Hmm. It looks like serial_ref is still in use for some aarch64 code, see lib.rs:488.

Yes, `serial_ref` and `mmio_transport_ref` are still in use, but the 3 functions above them (`boot_timer_ref`, `i8042_device_ref` and `rtc_device_ref`) aren't, so we can remove them to avoid carrying around dead code.

> So although currently unused, should we not keep these functions available so that ARM machines can still access these devices? They could be marked as aarch64 only.

I'm not sure I understand what you mean by this :(

Reply:
> Yes, serial_ref and mmio_transport_ref are still in use, but the 3 functions above them (boot_timer_ref, i8042_device_ref and rtc_device_ref) aren't, so we can remove them to avoid carrying around dead code.

* To me, it looks like we can get rid of `boot_timer_ref`. There doesn't seem to be a case where we would need a reference to the device after it has been added to the MMIO bus. It writes directly to the logs after boot and then we never need it again. 
* It's a similar story for the `RTCDevice` and `rtc_device_ref`.
* Right now, the I8042 device only seems to be used on x86_64, where it can be accessed through the `PortIODeviceManager`. However, whilst not in use, the device is also available to ARM64 devices. The way that this would be accessed would be using the MMIO bus, and so surely if we support it on ARM, we should keep a method around to access the device?

I am of the opinion that if the `I8042Device` is only going to be used on x86_64, we should tag it as such so that it is clear that this feature is not available to ARM. If there is a chance it is used on ARM systems in the future, and the way that it would be accessed is through the MMIO bus, then I don't know if it makes sense to remove the small function used to access it. Instead we could just tag it as `aarch64` only, so that it is clear that this function is only used to access the i8042 device on ARM systems.

Reply:
Ah, right. You can tag the `i8042` device as `#[cfg(target_arch = "x86_64")]` if you want, it's only used for the CTRL+ALT+ESC = restart functionality on x86 (which is not supported on ARM). So I don't see why we would ever need this decide on ARM (and if we ever do, we can just readd these function - however, keep in mind that on x86_64, the device can be used without these functions, so why would we suddenly need them on ARM?)

Reply:
> So boot_timer_ref and i8042_device_ref are actually dead code. We should remove them from here as Patrick said. We could also mark the I8042Device as x86 only all over the place. We should actually mark the vmm::devices::legacy::i8042 module x86 only.

Yeah, its not even too big of a change:

<details>

```diff
diff --git a/src/vmm/src/devices/bus.rs b/src/vmm/src/devices/bus.rs
index 2b016d730..e22fccfdf 100644
--- a/src/vmm/src/devices/bus.rs
+++ b/src/vmm/src/devices/bus.rs
@@ -51,15 +51,19 @@ pub struct Bus {
 }
 
 use event_manager::{EventOps, Events, MutEventSubscriber};
+use crate::devices::legacy::SerialDevice;
 
 #[cfg(target_arch = "aarch64")]
 use super::legacy::RTCDevice;
-use super::legacy::{I8042Device, SerialDevice};
+
+#[cfg(target_arch = "x86_64")]
+use super::legacy::{I8042Device};
 use super::pseudo::BootTimer;
 use super::virtio::mmio::MmioTransport;
 
 #[derive(Debug)]
 pub enum BusDevice {
+    #[cfg(target_arch = "x86_64")]
     I8042Device(I8042Device),
     #[cfg(target_arch = "aarch64")]
     RTCDevice(RTCDevice),
@@ -102,12 +106,6 @@ impl ConstantDevice {
 }
 
 impl BusDevice {
-    pub fn i8042_device_ref(&self) -> Option<&I8042Device> {
-        match self {
-            Self::I8042Device(x) => Some(x),
-            _ => None,
-        }
-    }
     #[cfg(target_arch = "aarch64")]
     pub fn rtc_device_ref(&self) -> Option<&RTCDevice> {
         match self {
@@ -133,13 +131,6 @@ impl BusDevice {
             _ => None,
         }
     }
-
-    pub fn i8042_device_mut(&mut self) -> Option<&mut I8042Device> {
-        match self {
-            Self::I8042Device(x) => Some(x),
-            _ => None,
-        }
-    }
     #[cfg(target_arch = "aarch64")]
     pub fn rtc_device_mut(&mut self) -> Option<&mut RTCDevice> {
         match self {
@@ -168,6 +159,7 @@ impl BusDevice {
 
     pub fn read(&mut self, offset: u64, data: &mut [u8]) {
         match self {
+            #[cfg(target_arch = "x86_64")]
             Self::I8042Device(x) => x.bus_read(offset, data),
             #[cfg(target_arch = "aarch64")]
             Self::RTCDevice(x) => x.bus_read(offset, data),
@@ -183,6 +175,7 @@ impl BusDevice {
 
     pub fn write(&mut self, offset: u64, data: &[u8]) {
         match self {
+            #[cfg(target_arch = "x86_64")]
             Self::I8042Device(x) => x.bus_write(offset, data),
             #[cfg(target_arch = "aarch64")]
             Self::RTCDevice(x) => x.bus_write(offset, data),
diff --git a/src/vmm/src/devices/legacy/mod.rs b/src/vmm/src/devices/legacy/mod.rs
index 224f77c7b..4642029d1 100644
--- a/src/vmm/src/devices/legacy/mod.rs
+++ b/src/vmm/src/devices/legacy/mod.rs
@@ -6,6 +6,7 @@
 // found in the THIRD-PARTY file.
 
 //! Implements legacy devices (UART, RTC etc).
+#[cfg(target_arch = "x86_64")]
 mod i8042;
 #[cfg(target_arch = "aarch64")]
 pub mod rtc_pl031;
@@ -19,6 +20,7 @@ use serde::Serializer;
 use utils::eventfd::EventFd;
 use vm_superio::Trigger;
 
+#[cfg(target_arch = "x86_64")]
 pub use self::i8042::{I8042Device, I8042Error as I8042DeviceError};
 #[cfg(target_arch = "aarch64")]
 pub use self::rtc_pl031::RTCDevice;
@@ -67,6 +69,7 @@ impl EventFdTrigger {
 /// Called by METRICS.flush(), this function facilitates serialization of aggregated metrics.
 pub fn flush_metrics<S: Serializer>(serializer: S) -> Result<S::Ok, S::Error> {
     let mut seq = serializer.serialize_map(Some(1))?;
+    #[cfg(target_arch = "x86_64")]
     seq.serialize_entry("i8042", &i8042::METRICS)?;
     #[cfg(target_arch = "aarch64")]
     seq.serialize_entry("rtc", &rtc_pl031::METRICS)?;
diff --git a/src/vmm/src/lib.rs b/src/vmm/src/lib.rs
index ee27194e4..ae1ef1bb1 100644
--- a/src/vmm/src/lib.rs
+++ b/src/vmm/src/lib.rs
@@ -213,6 +213,7 @@ pub enum VmmError {
     /// Event fd error: {0}
     EventFd(io::Error),
     /// I8042 error: {0}
+    #[cfg(target_arch = "x86_64")]
     I8042Error(devices::legacy::I8042DeviceError),
     /// Cannot access kernel file: {0}
     KernelFile(io::Error),
```

</details>

> However, maybe not related exactly with this PR, but could we actually get rid off all this mess, by turning BusDevice type to something like:

It actually used to be a trait: https://github.com/firecracker-microvm/firecracker/commit/0aecb443b8fd23673e37cb9dba82fe2ba5c290b6#diff-a3fcbea0daddf47c324fcbb2ac6738976593ec8f1961a9d7205088b434f1c537